File: cpp/graphAlgorithms/misc/toposort.cpp
/** @file toposort.cpp
 * 
 *  @author Jon Turner
 *  @date 2011
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

#include "stdinc.h"
#include "Adt.h"
#include "List.h"
#include "Digraph.h"

using namespace grafalgo;

/** Compute a topological ordering of dg.
 *  @param dg is a reference to an acyclic digraph
 *  @param vlist is a List object in which result is returned;
 *  on return, the vertices in vlist are in topogical order
 */
bool toposort(const Digraph& dg, List& vlist) {
	List q(dg.n());
	int *nin = new int[dg.n()+1];

	if (vlist.n() < dg.n()) vlist.resize(dg.n());
	else vlist.clear();

	// Let nin[u]=in-degree of u and put nodes u with nin[u]=0 on q
	for (vertex u = 1; u <= dg.n(); u++) {
		nin[u] = 0;
		for (edge e = dg.firstIn(u); e != 0; e = dg.nextIn(u,e)) {
			nin[u]++;
		}
		if (nin[u] == 0) q.addLast(u);
	}
	int i = 0;
	while (!q.empty()) { // q contains nodes u with nin[u] == 0
		vertex u = q.first(); q.removeFirst(); vlist.addLast(u); i++;
		for (edge e = dg.firstOut(u); e != 0; e = dg.nextOut(u,e)) {
			vertex v = dg.head(e);
			if ((--(nin[v])) == 0) q.addLast(v);
		}
	}
	return (i == dg.n());
}

File: js/graphAlgorithms/misc/toposort.mjs
/** @file toposort.cpp
 * 
 *  @author Jon Turner
 *  @date 2011
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

import List from '../../dataStructures/basic/List.mjs';

/** Compute a topological ordering of a directed Graph.
 *  @param g is a reference to an acyclic digraph
 *  @return an array listing the vertices in topological
 *  order, or null if the graph has a cycle.
 */
export default function toposort(g) {
	let vlist = []; // list of vertices in topological order
	let q = new List(g.n);
		// list of vertices whose in-edges have all been touched
	let icount = new Array(g.n+1).fill(0);
		// icount[u] = inDegree(u) - (# of times in-edge of u has been touched)

	// Initialize icount[u] and put nodes with icount[u]=0 on q
	for (let u = 1; u <= g.n; u++) {
		icount[u] = g.inDegree(u);
		if (icount[u] == 0) q.enq(u);
	}
	while (!q.empty()) { // q contains nodes u with icount[u] == 0
		let u = q.deq(); vlist.push(u);
		for (let e = g.firstOut(u); e != 0; e = g.nextOut(u,e)) {
			let v = g.head(e); icount[v]--;
			if (icount[v] == 0) q.enq(v);
		}
	}
	return (vlist.length == g.n ? vlist : null);
}
